include/circt/Dialect/Calyx/CalyxLoweringUtils.h

```diff
@@ -89,41 +131,351 @@ class WhileOpInterface {
   T impl;
 };
 
-// Walks the control of this component, and appends source information for leaf
-// nodes. It also appends a position attribute that connects the source location
-// metadata to the corresponding control operation.
-WalkResult
-getCiderSourceLocationMetadata(calyx::ComponentOp component,
-                               SmallVectorImpl<Attribute> &sourceLocations);
+//===----------------------------------------------------------------------===//
+// Lowering state classes
+//===----------------------------------------------------------------------===//
 
-// Tries to match a constant value defined by op. If the match was
-// successful, returns true and binds the constant to 'value'. If unsuccessful,
-// the value is unmodified.
-bool matchConstantOp(Operation *op, APInt &value);
+// An interface for the Calyx component lowering state. The `Loop` type must
+// implement the LoopInterface; it will be used for several lowering patterns.
+template <typename Loop>
+class ComponentLoweringStateInterface {
+  static_assert(std::is_base_of_v<LoopInterface, Loop>);
 
-// Returns true if there exists only a single memref::LoadOp which loads from
-// the memory referenced by loadOp.
-bool singleLoadFromMemory(Value memoryReference);
+public:
+  ComponentLoweringStateInterface(calyx::ComponentOp component)
+      : component(component) {}
 
-// Returns true if there are no memref::StoreOp uses with the referenced
-// memory.
-bool noStoresToMemory(Value memoryReference);
+  virtual ~ComponentLoweringStateInterface() = default;
```

Comment:
Should be moved to `CalyxLoweringUtils.cpp`, see [this](https://llvm.org/docs/CodingStandards.html#provide-a-virtual-method-anchor-for-classes-in-headers).

Reply:
Interesting. Thanks for pointing this out.